### ISSUE TITLE
Twice memory usage in `encode_multipart_formdata`

### ISSUE THREAD

**Issue by Axik** [shazow/urllib3#624]
Basically, if you wants to upload file like 1GB it will cost you 2GB RAM.

```
Filename: /Users/kyrylo/.virtualenvs/drsdk/lib/python2.7/site-packages/requests/packages/urllib3/filepost.py

Line #    Mem usage    Increment   Line Contents
================================================
    59    981.3 MiB      0.0 MiB   @profile
    60                             def encode_multipart_formdata(fields, boundary=None):
    61                                 """
    62                                 Encode a dictionary of ``fields`` using the multipart/form-data MIME format.
    63                             
    64                                 :param fields:
    65                                     Dictionary of fields or list of (key, :class:`~urllib3.fields.RequestField`).
    66                             
    67                                 :param boundary:
    68                                     If not specified, then a random boundary will be generated using
    69                                     :func:`mimetools.choose_boundary`.
    70                                 """
    71    981.3 MiB      0.0 MiB       body = BytesIO()
    72    981.3 MiB      0.0 MiB       if boundary is None:
    73    981.3 MiB      0.0 MiB           boundary = choose_boundary()
    74                             
    75   1920.7 MiB    939.4 MiB       for field in iter_field_objects(fields):
    76    981.3 MiB   -939.4 MiB           body.write(b('--%s\r\n' % (boundary)))    77                             
    78    981.3 MiB      0.0 MiB           writer(body).write(field.render_headers())
    79    981.3 MiB      0.0 MiB           data = field.data
    80                             
    81    981.3 MiB      0.0 MiB           if isinstance(data, int):
    82                                         data = str(data)  # Backwards compatibility
    83                             
    84    981.3 MiB      0.0 MiB           if isinstance(data, six.text_type):
    85                                         writer(body).write(data)
    86                                     else:
    87   1920.7 MiB    939.4 MiB               body.write(data)
    88                             
    89   1920.7 MiB      0.0 MiB           body.write(b'\r\n')
    90                             
    91   1920.7 MiB      0.0 MiB       body.write(b('--%s--\r\n' % (boundary)))
    92                             
    93   1920.7 MiB      0.0 MiB       content_type = str('multipart/form-data; boundary=%s' % boundary)
    94                             
    95   2000.1 MiB     79.4 MiB       return body.getvalue(), content_type
```

P.S. using this profiler https://pypi.python.org/pypi/memory_profiler

**Comment by shazow** [comment]
Right, 1x to read the data and 1x to store our encoded copy of it.

We had some discussions about converting this into a generator which would help bring it down to 1x unless you can provide the size upfront. I think there's a prototype or two somewhere, but it was never finished.

Is this something you'd be interested in working on?

**Comment by Axik** [comment]
```
Filename: /Users/kyrylo/.virtualenvs/drsdk/lib/python2.7/site-packages/requests/packages/urllib3/filepost.py

Line #    Mem usage    Increment   Line Contents
================================================
    59    284.1 MiB      0.0 MiB   @profile
    60                             def encode_multipart_formdata(fields, boundary=None):
    61                                 """
    62                                 Encode a dictionary of ``fields`` using the multipart/form-data MIME format.
    63                             
    64                                 :param fields:
    65                                     Dictionary of fields or list of (key, :class:`~urllib3.fields.RequestField`).
    66                             
    67                                 :param boundary:
    68                                     If not specified, then a random boundary will be generated using
    69                                     :func:`mimetools.choose_boundary`.
    70                                 """
    71    284.1 MiB      0.0 MiB       body = BytesIO()
    72    284.1 MiB      0.0 MiB       if boundary is None:
    73    284.2 MiB      0.1 MiB           boundary = choose_boundary()
    74                             
    75   1751.8 MiB   1467.6 MiB       for field in iter_field_objects(fields):
    76    284.2 MiB  -1467.6 MiB           body.write(b('--%s\r\n' % (boundary)))
    77                             
    78    284.2 MiB      0.0 MiB           writer(body).write(field.render_headers())
    79    284.2 MiB      0.0 MiB           data = field.data
    80                             
    81    284.2 MiB      0.0 MiB           if isinstance(data, int):
    82                                         data = str(data)  # Backwards compatibility
    83                             
    84    284.2 MiB      0.0 MiB           if isinstance(data, six.text_type):
    85                                         writer(body).write(data)
    86                                     else:
    87   1751.8 MiB   1467.6 MiB               body.write(data)
    88                             
    89   1751.8 MiB      0.0 MiB           body.write(b'\r\n')
    90                             
    91   1751.8 MiB      0.0 MiB       body.write(b('--%s--\r\n' % (boundary)))
    92                             
    93   1751.8 MiB      0.0 MiB       content_type = str('multipart/form-data; boundary=%s' % boundary)
    94                             
    95   2402.7 MiB    650.9 MiB       return body.getvalue(), content_type
```

In "cold" start state, it even triple it.

Also, what you think about not storing encoded copy, but making changes in-place, using mutable container.

**Comment by Lukasa** [comment]
note that the requests toolbelt has a [Streaming Multipart Data Encoder](http://toolbelt.readthedocs.org/en/latest/uploading-data.html#streaming-multipart-data-encoder) that may be interesting here.

**Comment by Axik** [comment]
thanks will try it just now

**Comment by sigmavirus24** [comment]
I really should make that requests agnostic and land it in urllib3. I think even urllib3 users would find it beneficial and it already heavily leverages some of urllib3's internals

**Comment by Axik** [comment]
@Lukasa it works.
Thank you.

@sigmavirus24 , @shazow 
Should I make PR to change behaviour to using this MultiPart Encoder if fileobject was provided?

**Comment by sigmavirus24** [comment]
@Axik that would take a great deal more work than you might realize. Also, the current default behaviour suits most users perfectly well so I don't think it's necessary

**Comment by shazow** [comment]
+100 to bringing the streaming encoder into urllib3.

**Comment by shazow** [comment]
(Related to #51, still would love to have this.)

**Comment by SethMichaelLarson** [comment]
FYI I'm working on this right now. Will submit a PR once I get a bigger chunk existing of tests passing.

**Comment by sigmavirus24** [comment]
@SethMichaelLarson what exactly are you doing?

**Comment by SethMichaelLarson** [comment]
Taking the old branch that was mentioned in #51, bringing it up to date with the module, adding Python 2 and 3 compatibility, fix all tests, add new tests. Currently have all but 4 tests current passing on the branch.

**Comment by sigmavirus24** [comment]
@SethMichaelLarson the resolution was to bring the code from requests-toolbelt into urllib3.  Please re-read the issue carefully.

**Comment by SethMichaelLarson** [comment]
@sigmavirus24 I thought I read the issue closely but I will re-read. The branch defines a `MultipartEncoderGenerator` object that is used in the same way that the tool-belt code would be used, I'll submit a PR today to show the work being done if you would like?

**Comment by pquentin** [comment]
@sethmlarson Can you please clarify why this is a breaking change? https://github.com/urllib3/urllib3/pull/1030 mentions that it should be deferred to v2, but I don't understand why.

**Comment by sethmlarson** [comment]
@pquentin I guess depending on how we accomplish this one it doesn't necessarily have to be a breaking change. We can deprecate `encode_multipart_formdata` in favor of a function that produces a `BytesIO` compatible body that can be `.tell()` for content-length and `.seek()` for rewinding.

I think the original thought was to keep the same function and change the return value hence the breaking change but we don't *have* to do that?

**Comment by Bobronium** [comment]
I'd like to work on this. 

I'll start from defining the scope and coming up with the strategy today. 

Looks like there were multiple attempts to address this. @sethmlarson, could you give some insights on why #1030 was closed?

**Comment by pquentin** [comment]
It was closed because we wanted to save it for the [v2 branch](https://github.com/urllib3/urllib3/tree/v2) that has died five years ago now, not because of anything specific in #1030 itself.

**Comment by pquentin** [comment]
However the preferred path now is getting https://github.com/urllib3/urllib3/pull/2331 merged. It's very close but not quite there yet. I think @sigmavirus24 can't accept the bounty, so we might be okay to award a part of it to whoever gets that PR merged.

**Comment by pquentin** [comment]
Please open a new pull request with "Closes #2331, Cloess #624" in the description.

I'm not 100% sure about the bounty, so if that's an issue you may prefer to wait for @sigmavirus24 and @sethmlarson to confirm.

**Comment by Bobronium** [comment]
Getting bounty is certainly desirable :)

**Comment by spacether** [comment]
@sethmlarson what is the priority on merging #2331? Why was it not included in v2?

**Comment by abebeos** [comment]
@sethmlarson , a general question:

Independent of that https://github.com/urllib3/urllib3/pull/2331 seemingly got stuck, would any other (possibly low code) solution be accepted (for the bounty), with the criteria that the memory usage halves?

**Comment by sethmlarson** [comment]
@abebeos You can create an independent PR, I added additional criteria to this issue for designing the API and implementing tests.

**Comment by abebeos** [comment]
The requirements you posted would fit with

- https://github.com/urllib3/urllib3/pull/2088 ?

What I meant was: fixing just the memory consumption, whilst leaving functionality intact. 

Or is "fixing only memory consumption" alone not of priority?

**Comment by alexwlchan** [comment]
Just sticking a note on this ticket: I'm gradually working on this in #3361, taking Ian's original PR and trying to get it into a mergeable state. I can only chip away at it slowly, but I'm getting there!

**Comment by zhangjiu513** [comment]
I’ve made considerable local progress on an independent implementation of the requirements in this issue, based on current `main`:

- added `urllib3.multipart.Part`, `MultipartEncoder`, and `MultipartDecoder`
- `Part` provides buffered `read()` / `peek()` and rewind support
- `MultipartEncoder.seek(0, 0)` attempts to rewind every part
- file-like field bodies stream in caller-sized chunks
- `encode_multipart_formdata()` now uses the encoder internally while preserving the existing output
- known multipart lengths are exposed to request framing as `Content-Length`
- added API/user documentation, unit tests, decoder round-trip tests, and a real local HTTP-server upload test

Local verification:
- 739 passed, 6 skipped in the relevant multipart/filepost/fields/request/connection-pool set
- Black, focused Ruff checks, and Mypy pass
- a 50 MiB file read in 64 KiB chunks completed in 801 reads with 262,396 bytes peak additional allocation under `tracemalloc`
- a deterministic test rejects unbounded reads

I’m on Windows, where urllib3’s nox configuration disables memray, so I have not claimed Linux memray verification. The full local suite also exposed unrelated failures caused by using PyPI Hypercorn 0.18 rather than urllib3’s pinned Hypercorn fork; the multipart-related tests pass.

I’m aware of the existing related PRs and have kept this implementation independent. Before opening another PR, would this API direction and verification approach be useful to maintainers, or would you prefer the work to be adapted to an existing effort?

**Comment by Ariyachan** [comment]
Would you consider a separately scoped, paid test-and-integration contribution toward #624, working with the existing implementation rather than adding another multipart API?

After reading #3361 and this issue, the potential scope I would propose is regression coverage for encoder rewind/partial reads and a reproducible memray check against the implementation the maintainers want to land. Before starting, we would need to agree which gaps remain, the target branch, and acceptance criteria.

Would a $100 portion of the bounty be available for that agreed scope, payable after the contribution is accepted? Could an approved invoice be paid through Open Collective to a PayPal account registered in mainland China? I would provide payment details privately through the payment process.

This is a scope and payment inquiry, not a claim on the issue or a report of completed work. If this work is already covered by the existing contributors, no duplicate effort is needed.

**Comment by sigmavirus24** [comment]
No bounties are currently being rewarded
